doc/cloud-init.md

```diff
@@ -241,6 +248,32 @@ config:
       - name: documentation_example
 ```
 
+(cloud-init-lxd-agent)=
+#### Enable the LXD agent
+
+The {ref}`VM images provided by LXD <remote-image-servers>` typically have the LXD agent enabled.
+However, if you use an image or ISO that does not include the LXD agent to create your VM, you can enable the LXD agent as follows:
+
+```yaml
+config:
+  cloud-init.user-data: |
+    #cloud-config
+    runcmd:
+      - mount -t 9p config /mnt
+      - cd /mnt
+      - ./install.sh
+      - cd /
+      - umount /mnt
+      - systemctl start lxd-agent
+```
+
+```{note}
+For these commands to work, the VM must be running a Linux distribution that uses `systemd`.
+
+Depending on the version of `cloud-init` that the distribution has, you might need to add the `config` disk device explicitly.
```

Comment:
@simondeziel Did I get this right?
I didn't manage to figure out how to test it (without trying the full ISO setup :wink: ).

Reply:
@ru-fu, yes, you've got this right and here's the not obvious way to quickly test without going over a manual ISO setup.

Create an instance and seed your SSH key:

```
$ lxc launch ubuntu-minimal-daily:24.04 --vm t1
$ lxc exec --user 1000 --group 1000 --env HOME=/home/ubuntu t1 -- ssh-import-id lp:sdeziel
2024-05-28 16:16:31,915 INFO Authorized key ['256', 'SHA256:BZU+QQkkZd5eveGqM1kMTGwl2xaaqTAmFMVX85/C2b4', 'simon@simon-laptop', '(ED25519)']
2024-05-28 16:16:31,916 INFO [1] SSH keys [Authorized]
```

Nuke the existing agent without using the agent (chicken/egg):

```
$ ssh -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no ubuntu@$(lxc list -c 4 -f csv t1 | cut -d\  -f1) sudo apt-get autopurge -y lxd-agent-loader
$ ssh -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no ubuntu@$(lxc list -c 4 -f csv t1 | cut -d\  -f1) sudo umount /run/lxd_agent
```

Confirm the agent no longer works:

```
$ ! lxc exec t1 -- true || false
Error: LXD VM agent isn't currently running
```

Manually install the agent:

```
$ ssh -t -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no ubuntu@$(lxc list -c 4 -f csv t1 | cut -d\  -f1) sudo /bin/bash << EOF
mount -t 9p config /mnt
cd /mnt
./install.sh
cd /
umount /mnt
systemctl start lxd-agent
EOF
```

Confirm it works:

```
$ lxc exec t1 -- hostname
t1
```

Reply:
@simondeziel Wow, yes, that's definitely something I wouldn't have come up with ... :laughing: 

That helped in actually testing the instructions - but unfortunately, it doesn't seem to work through cloud-init. :(

Here's what I did:
- Nuke the agent as you described above.
- Publish and export an image from the instance.
- Init a new VM, which doesn't have lxd-agent, from that image.
- Configure that image with the cloud-init user data as documented:

  ```
  $ lxc config show no-agent
  architecture: x86_64
  config:
    cloud-init.user-data: |
      #cloud-config
      runcmd:
        - mount -t 9p config /mnt
        - cd /mnt
        - ./install.sh
        - cd /
        - umount /mnt
        - systemctl start lxd-agent
    image.architecture: x86_64
  ...
  ```
- Start the instance and wait.
- Check if the agent is running ... it's not. :disappointed: 

I then fixed the agent as you explained above, checked the `cloud-init` status, and got:

```
root@no-agent:~# cloud-init status --long
status: error
extended_status: error - done
boot_status_code: enabled-by-generator
last_update: Wed, 29 May 2024 08:10:58 +0000
detail: DataSourceLXD
errors:
	- ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
recoverable_errors:
WARNING:
	- failed stage init-local
```

Any ideas? Did I miss any steps?

Reply:
Your testing procedure is sound. The error with the `DataSourceLXD` from `cloud-init` seems to be that `cloud-init` expects the data source to be working which depends on the `lxd-agent` to be functional.

I have an iso already imported here. I'll test it without any shortcuts to be sure what we have in the docs works. Thanks for checking!

Reply:
I finally got around to testing it and while it could work, it turns out to not be practical. Fixed/reworked instructions in https://github.com/canonical/lxd/pull/13568